<a href="https://colab.research.google.com/github/jonathanastudillo-lab/Ejercicio/blob/main/house_prices_advanced_regression_techniques_sumativo1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# House Prices: análisis exploratorio, limpieza de datos y métodos de separación de muestras

---

## Carrera: Ingeniería Civil Industrial
## Asignatura: Machine Learning

**Profesor:** Franco Andrés Mansilla

**Estudiantes:**  
- Jonathan Astudillo Vergara
- Gustavo ​Gutiérrez Díaz
- Mario Miño Farfan
- Robinson ​Sánchez Águila
- Leonel ​Tapia Araya  

**Fecha:** 29.03.2026

# Introducción

El presente trabajo tiene como objetivo desarrollar una primera aproximación al proceso de preparación de datos para un problema de machine learning, utilizando herramientas de análisis en Python a través de Google Colab. En particular, el desarrollo se enfoca en comprender y aplicar etapas fundamentales previas al modelamiento, tales como la exploración inicial de la base de datos, la limpieza de la información, el tratamiento de valores atípicos, la corrección de valores faltantes y la definición de estrategias de separación de muestras.

Para este trabajo se utilizó la base de datos **House Prices**, disponible en Kaggle, la cual contiene información de viviendas residenciales e incorpora múltiples variables relacionadas con sus características físicas, funcionales y de ubicación. El objetivo principal de este conjunto de datos es predecir el precio de venta de las viviendas (`SalePrice`). Debido a la diversidad de variables presentes en la base, esta resulta adecuada para aplicar distintas técnicas de análisis exploratorio y preparación de datos.

En una primera etapa, se revisan algunos frameworks utilizados en analítica de datos y machine learning, describiendo sus principales características, limitaciones y alcances, con el fin de justificar cuál resulta más adecuado para este caso. Posteriormente, se realiza el análisis de calidad de los datos, incluyendo estadística descriptiva, identificación y tratamiento de datos atípicos, revisión de valores faltantes y análisis de correlación entre variables. Finalmente, se presentan distintos métodos de separación de datos en muestras de entrenamiento, validación y test, con el propósito de seleccionar una estrategia apropiada para las siguientes etapas de modelamiento.

De esta forma, el trabajo busca no solo aplicar código en Python, sino también interpretar los resultados obtenidos y fundamentar cada una de las decisiones tomadas durante el proceso de preparación de la base de datos.

# Carga de librerías y base de datos

En esta sección se importan las librerías necesarias para trabajar con la base de datos, principalmente `pandas` y `numpy`, además de `ssl` para facilitar la carga del archivo desde una URL externa. Posteriormente, se carga la base de datos House Prices y se revisa su dimensión inicial.

In [1]:
import ssl
import pandas as pd
import numpy as np


ssl._create_default_https_context = ssl._create_unverified_context
# Usamos la URL raw para obtener solo el contenido del CSV
url = "https://raw.githubusercontent.com/jonathanastudillo-lab/Ejercicio/main/train_houseprice.csv"
df = pd.read_csv(url)
print(df.shape)
df.head()

(1460, 81)


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


La base de datos cargada corresponde al conjunto **House Prices** y presenta una dimensión de **1460 filas y 81 columnas**. Esto indica que se dispone de 1460 observaciones de viviendas y 81 variables asociadas a sus características físicas, funcionales y de venta.

La revisión de las primeras filas permite confirmar que la base contiene tanto variables numéricas como categóricas, lo que anticipa la necesidad de aplicar tratamientos diferenciados en las siguientes etapas del análisis.

# 1. Frameworks para aplicar analítica de datos

Un framework corresponde a un conjunto estructurado de herramientas, librerías y componentes que facilitan el desarrollo de soluciones de software o de analítica de datos. En machine learning, los frameworks permiten organizar el trabajo de preparación de datos, entrenamiento de modelos, validación y evaluación de resultados.

A continuación, se presentan tres frameworks ampliamente utilizados en analítica de datos y aprendizaje automático.


## 1.1 Scikit-learn

**Descripción:**  
Scikit-learn es un framework orientado al aprendizaje automático clásico en Python. Incluye herramientas para clasificación, regresión, clustering, reducción de dimensionalidad, preprocesamiento y evaluación de modelos.

**Limitaciones:**  
Una de sus principales limitaciones es que no está diseñado específicamente para redes neuronales profundas o modelos de deep learning de gran escala. Además, en flujos muy complejos puede requerir combinarse con otras librerías.

**Alcances:**  
Es especialmente útil para problemas tabulares, análisis exploratorio, pipelines de preprocesamiento, validación cruzada y modelos de machine learning supervisado y no supervisado.

---

## 1.2 TensorFlow / Keras

**Descripción:**  
TensorFlow es un framework orientado al desarrollo de modelos de aprendizaje profundo. Keras actúa como una interfaz de más alto nivel, facilitando la construcción, entrenamiento y evaluación de redes neuronales.

**Limitaciones:**  
Puede resultar más complejo que otras alternativas cuando el problema no requiere deep learning. Además, para tareas simples sobre datos tabulares puede ser innecesariamente pesado.

**Alcances:**  
Es adecuado para redes neuronales, procesamiento de imágenes, series de tiempo, procesamiento de lenguaje natural y problemas donde se requiere alta flexibilidad en arquitecturas de modelos.

---

## 1.3 PyTorch

**Descripción:**  
PyTorch es un framework de aprendizaje profundo ampliamente utilizado en investigación y aplicaciones avanzadas de inteligencia artificial. Se caracteriza por su flexibilidad y por facilitar la construcción de modelos complejos.

**Limitaciones:**  
Al igual que TensorFlow, puede ser excesivo para tareas simples de analítica de datos tabulares. También requiere mayor conocimiento técnico para aprovecharlo al máximo.

**Alcances:**  
Es especialmente útil en investigación, prototipado rápido de modelos, visión computacional, procesamiento de lenguaje natural y redes neuronales personalizadas.

---

## 1.4 Comparación y framework seleccionado

Los tres frameworks revisados son válidos, pero sus usos dependen del tipo de problema. Scikit-learn destaca por su simplicidad, rapidez y orientación al machine learning clásico, especialmente en datos tabulares. TensorFlow/Keras y PyTorch presentan mayores capacidades para deep learning, aunque también implican mayor complejidad.

Considerando que la base House Prices corresponde a un problema de regresión con datos tabulares estructurados, el framework más adecuado para este trabajo es **Scikit-learn**, ya que permite desarrollar de manera clara las etapas de separación de datos, validación y evaluación inicial del modelo.

# 2. Calidad de los datos

La calidad de los datos es una etapa fundamental en cualquier proyecto de machine learning, ya que la confiabilidad de los resultados depende directamente de la consistencia y preparación de la información de entrada. En esta sección se realiza una exploración inicial de la base House Prices y luego se aplican distintas técnicas de limpieza y preparación.

## 2.1 Exploración inicial y estadística descriptiva

En primer lugar, se revisa la dimensión de la base de datos y se observan sus primeras filas, con el objetivo de entender su estructura general. Luego, las variables se separan en numéricas y categóricas, permitiendo diferenciar su tratamiento analítico. Finalmente, se genera un resumen estadístico descriptivo de las variables numéricas, con medidas como media, desviación estándar, mínimo, máximo y percentiles.

In [2]:
# Ver tamaño y primeras filas
print("Dimensión del DataFrame:", df.shape)
display(df.head())

# Separar variables numéricas y categóricas
num_cols = df.select_dtypes(include=[np.number]).columns.tolist()
cat_cols = df.select_dtypes(exclude=[np.number]).columns.tolist()

print("Cantidad de variables numéricas:", len(num_cols))
print("Cantidad de variables categóricas:", len(cat_cols))

print("\nVariables numéricas:")
print(num_cols)

print("\nVariables categóricas:")
print(cat_cols)

# Resumen estadístico de variables numéricas
resumen_numerico = df[num_cols].describe().T
display(resumen_numerico)

# Función para winsorizar entre percentil 1 y percentil 99
def winsorize_p1_p99(df, cols):
    df_out = df.copy()
    bounds = {}

    for c in cols:
        p1 = df_out[c].quantile(0.01)
        p99 = df_out[c].quantile(0.99)
        bounds[c] = (p1, p99)
        df_out[c] = df_out[c].clip(lower=p1, upper=p99)

    return df_out, bounds

Dimensión del DataFrame: (1460, 81)


,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65.0,8450,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2008,WD,Normal,208500
1,2,20,RL,80.0,9600,Pave,NaN,Reg,Lvl,AllPub,...,0,NaN,NaN,NaN,0,5,2007,WD,Normal,181500
2,3,60,RL,68.0,11250,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,9,2008,WD,Normal,223500
3,4,70,RL,60.0,9550,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84.0,14260,Pave,NaN,IR1,Lvl,AllPub,...,0,NaN,NaN,NaN,0,12,2008,WD,Normal,250000


Cantidad de variables numéricas: 38
Cantidad de variables categóricas: 43

Variables numéricas:
['Id', 'MSSubClass', 'LotFrontage', 'LotArea', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal', 'MoSold', 'YrSold', 'SalePrice']

Variables categóricas:
['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 

,count,mean,std,min,25%,50%,75%,max
Id,1460.0,730.500000,421.610009,1.0,365.75,730.5,1095.25,1460.0
MSSubClass,1460.0,56.897260,42.300571,20.0,20.00,50.0,70.00,190.0
LotFrontage,1201.0,70.049958,24.284752,21.0,59.00,69.0,80.00,313.0
LotArea,1460.0,10516.828082,9981.264932,1300.0,7553.50,9478.5,11601.50,215245.0
OverallQual,1460.0,6.099315,1.382997,1.0,5.00,6.0,7.00,10.0
OverallCond,1460.0,5.575342,1.112799,1.0,5.00,5.0,6.00,9.0
YearBuilt,1460.0,1971.267808,30.202904,1872.0,1954.00,1973.0,2000.00,2010.0
YearRemodAdd,1460.0,1984.865753,20.645407,1950.0,1967.00,1994.0,2004.00,2010.0
MasVnrArea,1452.0,103.685262,181.066207,0.0,0.00,0.0,166.00,1600.0
BsmtFinSF1,1460.0,443.639726,456.098091,0.0,0.00,383.5,712.25,5644.0


La separación entre variables numéricas y categóricas permite estructurar adecuadamente el análisis exploratorio. Las variables numéricas pueden ser evaluadas mediante estadísticas descriptivas y tratamiento de valores extremos, mientras que las categóricas requieren un tratamiento distinto, principalmente enfocado en frecuencia, imputación y posterior codificación.

El resumen estadístico muestra que existen variables con escalas muy diferentes, así como posibles valores extremos en atributos como `SalePrice`, `GrLivArea` y `LotFrontage`, lo que justifica aplicar un tratamiento para reducir el efecto de outliers.

## 2.2 Identificación y tratamiento de datos atípicos

Para reducir el efecto de valores extremos en las variables numéricas, se aplicó una estrategia de **winsorización** entre el percentil 1 y el percentil 99. Esta técnica consiste en reemplazar los valores inferiores al percentil 1 y superiores al percentil 99 por dichos límites, evitando eliminar observaciones y reduciendo la distorsión que los outliers pueden provocar en el análisis y en modelos posteriores.

Posteriormente, se comparan estadísticas antes y después de la winsorización para observar el efecto del tratamiento sobre variables como `SalePrice`, `GrLivArea` y `LotFrontage`.

In [3]:
def winsorize_p1_p99(df, cols):
    df_out = df.copy()
    bounds = {}

    for c in cols:
        p1 = df_out[c].quantile(0.01)
        p99 = df_out[c].quantile(0.99)
        bounds[c] = (p1, p99)
        df_out[c] = df_out[c].clip(lower=p1, upper=p99)

    return df_out, bounds

df_w, bounds = winsorize_p1_p99(df, num_cols)

bounds_df = pd.DataFrame(bounds, index=["p1", "p99"]).T
bounds_df.index.name = "Variable"

print("Percentiles usados para winsorización:")
display(bounds_df)

print("Resumen ANTES de winsorizar:")
display(df[num_cols].describe().T)

print("Resumen DESPUÉS de winsorizar:")
display(df_w[num_cols].describe().T)

Percentiles usados para winsorización:


,p1,p99
Variable,,
Id,15.59,1445.41
MSSubClass,20.00,190.00
LotFrontage,21.00,141.00
LotArea,1680.00,37567.64
OverallQual,3.00,10.00
OverallCond,3.00,9.00
YearBuilt,1899.18,2009.00
YearRemodAdd,1950.00,2009.00
MasVnrArea,0.00,791.92


Resumen ANTES de winsorizar:


,count,mean,std,min,25%,50%,75%,max
Id,1460.0,730.500000,421.610009,1.0,365.75,730.5,1095.25,1460.0
MSSubClass,1460.0,56.897260,42.300571,20.0,20.00,50.0,70.00,190.0
LotFrontage,1201.0,70.049958,24.284752,21.0,59.00,69.0,80.00,313.0
LotArea,1460.0,10516.828082,9981.264932,1300.0,7553.50,9478.5,11601.50,215245.0
OverallQual,1460.0,6.099315,1.382997,1.0,5.00,6.0,7.00,10.0
OverallCond,1460.0,5.575342,1.112799,1.0,5.00,5.0,6.00,9.0
YearBuilt,1460.0,1971.267808,30.202904,1872.0,1954.00,1973.0,2000.00,2010.0
YearRemodAdd,1460.0,1984.865753,20.645407,1950.0,1967.00,1994.0,2004.00,2010.0
MasVnrArea,1452.0,103.685262,181.066207,0.0,0.00,0.0,166.00,1600.0
BsmtFinSF1,1460.0,443.639726,456.098091,0.0,0.00,383.5,712.25,5644.0


Resumen DESPUÉS de winsorizar:


,count,mean,std,min,25%,50%,75%,max
Id,1460.0,730.500000,421.343430,15.59,365.75,730.5,1095.25,1445.41
MSSubClass,1460.0,56.897260,42.300571,20.00,20.00,50.0,70.00,190.00
LotFrontage,1201.0,69.600333,21.723531,21.00,59.00,69.0,80.00,141.00
LotArea,1460.0,10063.006575,5062.299090,1680.00,7553.50,9478.5,11601.50,37567.64
OverallQual,1460.0,6.104110,1.369444,3.00,5.00,6.0,7.00,10.00
OverallCond,1460.0,5.580137,1.098826,3.00,5.00,5.0,6.00,9.00
YearBuilt,1460.0,1971.413493,29.806332,1899.18,1954.00,1973.0,2000.00,2009.00
YearRemodAdd,1460.0,1984.861644,20.640499,1950.00,1967.00,1994.0,2004.00,2009.00
MasVnrArea,1452.0,101.174793,168.046772,0.00,0.00,0.0,166.00,791.92
BsmtFinSF1,1460.0,438.530240,429.049203,0.00,0.00,383.5,712.25,1572.41


### Análisis del resultado

La winsorización aplicada entre los percentiles 1 y 99 permitió limitar el efecto de valores extremos en las variables numéricas sin eliminar observaciones de la base. Esto es relevante porque, en lugar de perder datos, se conservan los registros, pero reduciendo la influencia de valores atípicos que podrían distorsionar medidas estadísticas y afectar el desempeño de modelos posteriores.

La comparación entre los resúmenes antes y después del tratamiento muestra que los máximos y mínimos extremos fueron ajustados a límites más representativos de la distribución general de los datos.

A continuación, se revisan algunas variables específicas para observar el efecto de la winsorización antes y después del tratamiento.


In [4]:
print("Variable SalePrice")
display(bounds_df.loc["SalePrice"])

print("Antes:")
print(df["SalePrice"].describe())

print("\nDespués:")
print(df_w["SalePrice"].describe())

print("\nVariable GrLivArea")
print("Antes:")
print(df["GrLivArea"].describe())

print("\nDespués:")
print(df_w["GrLivArea"].describe())

print("\nVariable LotFrontage")
print("Antes:")
print(df["LotFrontage"].describe())

print("\nDespués:")
print(df_w["LotFrontage"].describe())

Variable SalePrice


,SalePrice
p1,61815.97
p99,442567.01


Antes:
count      1460.000000
mean     180921.195890
std       79442.502883
min       34900.000000
25%      129975.000000
50%      163000.000000
75%      214000.000000
max      755000.000000
Name: SalePrice, dtype: float64

Después:
count      1460.000000
mean     179926.416233
std       74052.658892
min       61815.970000
25%      129975.000000
50%      163000.000000
75%      214000.000000
max      442567.010000
Name: SalePrice, dtype: float64

Variable GrLivArea
Antes:
count    1460.000000
mean     1515.463699
std       525.480383
min       334.000000
25%      1129.500000
50%      1464.000000
75%      1776.750000
max      5642.000000
Name: GrLivArea, dtype: float64

Después:
count    1460.000000
mean     1510.289658
std       495.148927
min       692.180000
25%      1129.500000
50%      1464.000000
75%      1776.750000
max      3123.480000
Name: GrLivArea, dtype: float64

Variable LotFrontage
Antes:
count    1201.000000
mean       70.049958
std        24.284752
min        21.000000
2

La revisión puntual de variables como `SalePrice`, `GrLivArea` y `LotFrontage` permite observar con mayor claridad el efecto de la winsorización. En estos casos, los valores extremos fueron acotados, reduciendo la dispersión excesiva en la cola superior de la distribución.

Esto es especialmente importante en `SalePrice`, ya que corresponde a la variable objetivo del problema de regresión. Un tratamiento adecuado de sus valores extremos ayuda a estabilizar el análisis y mejora la consistencia de etapas posteriores.

## 2.3 Identificación y tratamiento de missing values

Una vez tratados los valores extremos, se revisó la presencia de valores faltantes en la base de datos. Para ello, se calculó tanto el conteo absoluto como el porcentaje de missing values por variable.

El tratamiento se realizó de forma diferenciada según el significado de la ausencia de datos. En algunas variables, el valor faltante no necesariamente representa un error de registro, sino la ausencia real de una característica en la vivienda. Por ejemplo, si una casa no posee piscina, cerco, chimenea, garaje o sótano, es esperable que ciertas variables asociadas aparezcan vacías.

Por esta razón, se eliminaron algunas variables con exceso extremo de missing values, se imputó la categoría `"None"` en variables categóricas donde la ausencia representa que la característica no existe, se imputó 0 en variables numéricas asociadas a ausencia real, y en el resto de las variables se utilizaron la moda y la mediana como criterios de imputación. Además, en `LotFrontage` se aplicó una imputación basada en la mediana por barrio (`Neighborhood`), para mantener mayor coherencia con el contexto de la variable.

In [5]:
# Conteo de missing
missing_count = df_w.isna().sum().sort_values(ascending=False)
missing_pct = (df_w.isna().mean() * 100).sort_values(ascending=False)

pd.DataFrame({"missing_count": missing_count, "missing_pct": missing_pct}).head(20)

,missing_count,missing_pct
PoolQC,1453,99.520548
MiscFeature,1406,96.301370
Alley,1369,93.767123
Fence,1179,80.753425
MasVnrType,872,59.726027
FireplaceQu,690,47.260274
LotFrontage,259,17.739726
GarageQual,81,5.547945
GarageFinish,81,5.547945
GarageType,81,5.547945


### Análisis del resultado

El análisis de valores faltantes muestra que algunas variables presentan una proporción considerable de missing values. Sin embargo, en esta base de datos no todos los valores faltantes deben interpretarse como errores de registro, ya que en varios casos representan la ausencia real de una característica en la vivienda.

Esto ocurre, por ejemplo, en variables asociadas a piscina, cerco, chimenea, garaje o sótano. Por lo tanto, el tratamiento de missing values no puede hacerse de manera uniforme, sino que debe considerar el significado real de cada variable.


In [6]:
df_clean = df_w.copy()

# 1) Eliminar columnas con exceso extremo de missing
drop_cols = ["PoolQC", "MiscFeature", "Alley", "Fence"]
drop_cols = [c for c in drop_cols if c in df_clean.columns]
df_clean.drop(columns=drop_cols, inplace=True)

# 2) Categóricas donde el NA significa ausencia real
fill_none_cols = [
    "MasVnrType",
    "FireplaceQu",
    "GarageType", "GarageFinish", "GarageQual", "GarageCond",
    "BsmtQual", "BsmtCond", "BsmtExposure", "BsmtFinType1", "BsmtFinType2"
]

for c in fill_none_cols:
    if c in df_clean.columns:
        df_clean[c] = df_clean[c].fillna("None")

# 3) Numéricas donde el NA también representa ausencia real
fill_zero_cols = [
    "MasVnrArea",
    "BsmtFinSF1", "BsmtFinSF2", "BsmtUnfSF", "TotalBsmtSF",
    "BsmtFullBath", "BsmtHalfBath",
    "GarageCars", "GarageArea", "GarageYrBlt"
]

for c in fill_zero_cols:
    if c in df_clean.columns:
        df_clean[c] = df_clean[c].fillna(0)

# 4) LotFrontage: imputar con mediana por Neighborhood
if "LotFrontage" in df_clean.columns and "Neighborhood" in df_clean.columns:
    df_clean["LotFrontage"] = df_clean.groupby("Neighborhood")["LotFrontage"].transform(
        lambda x: x.fillna(x.median())
    )
    df_clean["LotFrontage"] = df_clean["LotFrontage"].fillna(df_clean["LotFrontage"].median())

# 5) Categóricas restantes: moda
cat_cols_clean = df_clean.select_dtypes(exclude=[np.number]).columns.tolist()
for c in cat_cols_clean:
    if df_clean[c].isna().sum() > 0:
        df_clean[c] = df_clean[c].fillna(df_clean[c].mode()[0])

# 6) Numéricas restantes: mediana
num_cols_clean = df_clean.select_dtypes(include=[np.number]).columns.tolist()
for c in num_cols_clean:
    if df_clean[c].isna().sum() > 0:
        df_clean[c] = df_clean[c].fillna(df_clean[c].median())

missing_final = df_clean.isna().sum().sort_values(ascending=False)
print(missing_final.head(20))
print("\nTotal de missing restantes:", df_clean.isna().sum().sum())
print("\nDimensión final:", df_clean.shape)

Id              0
MSSubClass      0
MSZoning        0
LotFrontage     0
LotArea         0
Street          0
LotShape        0
LandContour     0
Utilities       0
LotConfig       0
LandSlope       0
Neighborhood    0
Condition1      0
Condition2      0
BldgType        0
HouseStyle      0
OverallQual     0
OverallCond     0
YearBuilt       0
YearRemodAdd    0
dtype: int64

Total de missing restantes: 0

Dimensión final: (1460, 77)


Luego del tratamiento de missing values, la base quedó sin valores faltantes, lo que confirma que el proceso de limpieza fue exitoso. Además, la dimensión final refleja la eliminación de variables con exceso extremo de missing, mientras que el resto de las ausencias fue tratado mediante imputación coherente con el significado de cada atributo.

Este resultado es importante porque deja la base en condiciones adecuadas para avanzar hacia la etapa de modelamiento, evitando errores o sesgos asociados a datos incompletos.

## 2.4 Correlación entre variables

Una vez corregidos los valores faltantes, se evaluó la correlación entre variables numéricas mediante el coeficiente de Pearson. Este análisis permite detectar relaciones lineales fuertes y, en particular, identificar variables que podrían estar aportando información redundante.

La presencia de variables altamente correlacionadas puede dificultar la interpretación y generar problemas de multicolinealidad, especialmente en modelos lineales. A partir de este análisis, se revisaron los pares con mayor correlación absoluta y se decidió eliminar algunas variables redundantes, conservando aquellas con mayor sentido interpretativo o mejor relación con la variable objetivo `SalePrice`.

In [7]:
# Recalcular columnas numéricas desde df_clean
num_cols_clean = df_clean.select_dtypes(include=[np.number]).columns.tolist()

# Matriz de correlación
corr = df_clean[num_cols_clean].corr(method="pearson")

# Ver matriz completa
display(corr)

# Ver correlaciones más fuertes entre variables (sin repetir pares)
corr_abs = corr.abs()
upper = corr_abs.where(np.triu(np.ones(corr_abs.shape), k=1).astype(bool))

top_pairs = upper.stack().sort_values(ascending=False)
display(top_pairs.head(20))

,Id,MSSubClass,LotFrontage,LotArea,OverallQual,OverallCond,YearBuilt,YearRemodAdd,MasVnrArea,BsmtFinSF1,...,WoodDeckSF,OpenPorchSF,EnclosedPorch,3SsnPorch,ScreenPorch,PoolArea,MiscVal,MoSold,YrSold,SalePrice
Id,1.000000,0.011176,-0.015388,-0.006339,-0.029298,0.008899,-0.011157,-0.022063,-0.047871,-0.012354,...,-0.031724,-0.002958,0.005680,-0.040929,-0.001365,NaN,-0.034645,0.021203,0.000616,-0.028571
MSSubClass,0.011176,1.000000,-0.406366,-0.287184,0.030368,-0.062109,0.030109,0.040655,0.021414,-0.072091,...,-0.016440,-0.004478,-0.014892,-0.034923,-0.027292,NaN,-0.027886,-0.013585,-0.021407,-0.088827
LotFrontage,-0.015388,-0.406366,1.000000,0.510400,0.239809,-0.049222,0.129764,0.077974,0.205458,0.166207,...,0.101727,0.142559,-0.010138,0.065043,0.049834,NaN,0.017470,0.007872,0.006047,0.375113
LotArea,-0.006339,-0.287184,0.510400,1.000000,0.184171,-0.018027,0.040490,0.037817,0.160788,0.232365,...,0.222785,0.160257,-0.018566,0.051020,0.095568,NaN,0.060646,-0.005922,-0.035254,0.386883
OverallQual,-0.029298,0.030368,0.239809,0.184171,1.000000,-0.108943,0.582135,0.550096,0.420436,0.224470,...,0.242538,0.324401,-0.127305,0.029500,0.060444,NaN,-0.081100,0.067026,-0.027131,0.809971
OverallCond,0.008899,-0.062109,-0.049222,-0.018027,-0.108943,1.000000,-0.386027,0.070660,-0.133267,-0.045502,...,-0.002915,-0.045826,0.077420,0.039026,0.051329,NaN,0.086106,-0.003845,0.046993,-0.085864
YearBuilt,-0.011157,0.030109,0.129764,0.040490,0.582135,-0.386027,1.000000,0.601027,0.326042,0.251577,...,0.232431,0.220626,-0.388990,0.027767,-0.047156,NaN,-0.078449,0.014185,-0.012543,0.551515
YearRemodAdd,-0.022063,0.040655,0.077974,0.037817,0.550096,0.070660,0.601027,1.000000,0.183263,0.124823,...,0.215385,0.248363,-0.202792,0.051896,-0.036148,NaN,-0.071507,0.021617,0.035424,0.529343
MasVnrArea,-0.047871,0.021414,0.205458,0.160788,0.420436,-0.133267,0.326042,0.183263,1.000000,0.245058,...,0.170067,0.146520,-0.114207,0.026466,0.061617,NaN,-0.044255,-0.002836,-0.000103,0.471914
BsmtFinSF1,-0.012354,-0.072091,0.166207,0.232365,0.224470,-0.045502,0.251577,0.124823,0.245058,1.000000,...,0.207069,0.091508,-0.110666,0.041042,0.070040,NaN,0.002470,-0.003669,0.014157,0.394831


,,0
GarageCars,GarageArea,0.892964
GrLivArea,TotRmsAbvGrd,0.836481
OverallQual,SalePrice,0.809971
TotalBsmtSF,1stFlrSF,0.803791
GrLivArea,SalePrice,0.721656
2ndFlrSF,GrLivArea,0.689593
BedroomAbvGr,TotRmsAbvGrd,0.676355
GarageCars,SalePrice,0.665698
BsmtFinSF1,BsmtFullBath,0.664606
GrLivArea,FullBath,0.649000


El análisis de correlación permite identificar variables numéricas que presentan relaciones lineales fuertes entre sí. Esto es útil para detectar posibles redundancias de información, ya que dos variables altamente correlacionadas pueden estar describiendo prácticamente el mismo fenómeno.

Entre los pares con mayor correlación se observan relaciones como `GarageCars` con `GarageArea`, `GrLivArea` con `TotRmsAbvGrd`, y `TotalBsmtSF` con `1stFlrSF`. Estas relaciones justifican evaluar la eliminación de algunas variables para reducir multicolinealidad y simplificar el conjunto de predictores.

In [8]:
cols_redundantes = ["GarageArea", "TotRmsAbvGrd", "1stFlrSF"]

df_reduced = df_clean.drop(columns=cols_redundantes, errors="ignore")

print("Columnas eliminadas:", cols_redundantes)
print("Dimensión original:", df_clean.shape)
print("Dimensión reducida:", df_reduced.shape)

Columnas eliminadas: ['GarageArea', 'TotRmsAbvGrd', '1stFlrSF']
Dimensión original: (1460, 77)
Dimensión reducida: (1460, 74)


A partir del análisis de correlación, se eliminaron tres variables consideradas redundantes: `GarageArea`, `TotRmsAbvGrd` y `1stFlrSF`. La decisión se tomó conservando variables con mayor sentido interpretativo o mejor relación con la variable objetivo `SalePrice`.

Esta reducción permite simplificar la base de datos, disminuir el riesgo de multicolinealidad y mantener un conjunto de variables más consistente para las siguientes etapas del trabajo.

## 2.5 Síntesis de la calidad de datos

Como resultado del proceso de preparación, la base quedó sin valores faltantes, con tratamiento aplicado a los valores extremos y con una reducción de variables redundantes a partir del análisis de correlación. Esto permitió obtener una base más consistente y adecuada para continuar con la etapa de separación en muestras de entrenamiento, validación y test.

# 3. Ventanas de desarrollo, validación y test

La separación de datos es una etapa clave en machine learning, ya que permite evaluar la capacidad de generalización del modelo sobre datos no vistos. En esta sección se presentan tres métodos de partición de la muestra y luego se selecciona uno de ellos para su aplicación sobre la base House Prices.

## 3.1 Preparación de la base para modelar

Antes de aplicar los métodos de separación, se definió una base final de modelamiento. Para ello, se eliminó la variable `Id`, ya que corresponde a un identificador sin capacidad predictiva, y se definió `SalePrice` como variable objetivo. El resto de las columnas fueron consideradas como variables predictoras.

In [9]:
df_model = df_reduced.copy()

df_model = df_model.drop(columns=["Id"], errors="ignore")

X = df_model.drop("SalePrice", axis=1)
y = df_model["SalePrice"]

print("X:", X.shape)
print("y:", y.shape)

print("Columnas finales:", len(df_model.columns))
print("Variables predictoras:", X.shape[1])
print("Observaciones:", X.shape[0])

X: (1460, 72)
y: (1460,)
Columnas finales: 73
Variables predictoras: 72
Observaciones: 1460


La base final de modelamiento quedó estructurada con una variable objetivo (`SalePrice`) y un conjunto de variables predictoras que servirán de entrada para los distintos métodos de separación de muestras.

Además, se eliminó la variable `Id`, ya que corresponde únicamente a un identificador y no aporta información útil para la predicción. Con esto, la base queda formalmente preparada para avanzar hacia la etapa de partición en entrenamiento, validación y test.

## 3.2 Método 1: Train / Test

El método Train/Test divide la base de datos en dos subconjuntos: uno de entrenamiento y otro de prueba. Su principal ventaja es la simplicidad, ya que permite realizar una partición rápida y fácil de interpretar. Sin embargo, presenta la desventaja de no disponer de una muestra específica para validación, por lo que puede ser menos apropiado cuando se busca ajustar decisiones del modelo de manera más controlada.

In [10]:
from sklearn.model_selection import train_test_split

X_train_m1, X_test_m1, y_train_m1, y_test_m1 = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

print("X_train_m1:", X_train_m1.shape)
print("X_test_m1:", X_test_m1.shape)
print("y_train_m1:", y_train_m1.shape)
print("y_test_m1:", y_test_m1.shape)

X_train_m1: (1168, 72)
X_test_m1: (292, 72)
y_train_m1: (1168,)
y_test_m1: (292,)


La partición Train/Test distribuyó la muestra en un subconjunto de entrenamiento y otro de prueba, manteniendo la mayor parte de las observaciones para el aprendizaje del modelo. Esta estrategia resulta útil como aproximación inicial, aunque no incorpora una muestra independiente de validación para apoyar decisiones intermedias durante el desarrollo.

## 3.3 Método 2: Train / Validation / Test

Este método divide la muestra en tres subconjuntos: entrenamiento, validación y test. La muestra de entrenamiento se utiliza para que el modelo aprenda patrones, la muestra de validación permite ajustar decisiones durante el desarrollo, y la muestra de test queda reservada para la evaluación final sobre datos no vistos.

Su principal ventaja es que entrega una estructura más ordenada y coherente para el proceso de modelamiento. Como desventaja, cada subconjunto queda con menor cantidad de observaciones que en una división más simple.

In [11]:
from sklearn.model_selection import train_test_split

# Primer split: separar test
X_temp_m2, X_test_m2, y_temp_m2, y_test_m2 = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# Segundo split: separar train y validation
X_train_m2, X_val_m2, y_train_m2, y_val_m2 = train_test_split(
    X_temp_m2, y_temp_m2, test_size=0.25, random_state=42
)

print("X_train_m2:", X_train_m2.shape)
print("X_val_m2:", X_val_m2.shape)
print("X_test_m2:", X_test_m2.shape)

print("y_train_m2:", y_train_m2.shape)
print("y_val_m2:", y_val_m2.shape)
print("y_test_m2:", y_test_m2.shape)

X_train_m2: (876, 72)
X_val_m2: (292, 72)
X_test_m2: (292, 72)
y_train_m2: (876,)
y_val_m2: (292,)
y_test_m2: (292,)


La partición Train / Validation / Test generó tres subconjuntos diferenciados para entrenamiento, validación y prueba. Esta estructura permite separar de manera más clara el proceso de aprendizaje, el ajuste de decisiones y la evaluación final del modelo, por lo que resulta más adecuada para el desarrollo ordenado del trabajo.

## 3.4 Método 3: Cross Validation

La validación cruzada divide la base en varios subconjuntos o folds, utilizando algunos para entrenamiento y otro para validación en cada iteración. De esta forma, el desempeño del modelo no depende de una sola partición, sino que se evalúa en varias combinaciones.

Como ventaja, este método entrega una estimación más robusta del desempeño. Como desventaja, implica mayor costo computacional y una implementación más extensa. En este trabajo, se aplicó validación cruzada con 5 folds utilizando un modelo base de regresión lineal, previa codificación de variables categóricas.


In [12]:
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LinearRegression

X_encoded = pd.get_dummies(X, drop_first=True)

modelo = LinearRegression()

kf = KFold(n_splits=5, shuffle=True, random_state=42)

scores = cross_val_score(
    modelo,
    X_encoded,
    y,
    cv=kf,
    scoring="r2"
)

print("Scores por fold:", scores)
print("Promedio R2:", scores.mean())

Scores por fold: [0.84876511 0.92316718 0.77602107 0.84953853 0.8518279 ]
Promedio R2: 0.8498639579764931


En este caso, el promedio de `R²` obtenido fue de **0.85**, lo que indica que el modelo logra explicar aproximadamente el **85%** de la variabilidad del precio de venta. Los scores por fold muestran una variación moderada, lo que sugiere que el comportamiento del modelo es razonablemente consistente entre distintas particiones de la muestra.

## 3.5 Método seleccionado

De los tres métodos revisados, se seleccionó **Train / Validation / Test** como estrategia principal para este trabajo. La razón es que permite separar de forma clara las etapas de entrenamiento, validación y prueba final, alineándose además con lo solicitado en la pauta.

Si bien la validación cruzada entrega una evaluación más robusta del desempeño, el esquema Train / Validation / Test resulta más claro para organizar el proceso de desarrollo del modelo dentro del contexto del trabajo.

# Conclusión

En este trabajo se desarrolló un proceso inicial de preparación de datos aplicado a la base House Prices, abordando tanto aspectos conceptuales como prácticos del flujo de machine learning. En primer lugar, se revisaron distintos frameworks utilizados en analítica de datos, concluyendo que Scikit-learn resulta el más adecuado para este caso debido a la naturaleza tabular del problema y a su facilidad de uso en tareas de regresión, validación y evaluación.

Posteriormente, se realizó una exploración inicial de la base y un proceso de limpieza que incluyó estadística descriptiva, tratamiento de datos atípicos mediante winsorización, imputación de valores faltantes con criterios diferenciados según el significado de la ausencia y análisis de correlación para reducir redundancia entre variables.

Finalmente, se compararon tres métodos de separación de datos: Train/Test, Train/Validation/Test y Cross Validation. A partir de esta comparación, se seleccionó el método Train/Validation/Test, ya que permite organizar de mejor manera las etapas de entrenamiento, validación y prueba final, dejando la base preparada para futuras etapas de modelamiento.

# Bibliografía


- House prices - advanced regression techniques. (s. f.). Kaggle. https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques/data?  select=train.csv  
- Communications. (2026, 23 febrero). «Machine learning»: ¿qué es y cómo funciona el maestro en reconocer patrones? BBVA NOTICIAS. https://www.bbva.com/es/innovacion/machine-learning-que-es-y-como-funciona/  
- Infinite Codes. (2024, 17 septiembre). All Machine Learning algorithms explained in 17 min [Vídeo]. YouTube. https://www.youtube.com/watch?v=E0Hmnixke2g  
- Linkfy. (2021, 20 octubre). ¿Qué es machine learning? Aprendizaje automático [Vídeo]. YouTube. https://www.youtube.com/watch?v=xrQ1YH0PnrM  
- Google Colab. (s. f.). https://colab.research.google.com/notebooks/markdown_guide.ipynb

In [13]:
#actualización 19/04/2026; se genera CSV de la base datos limpia para continuar al trabajo sumativo 2
df_model.to_csv("house_prices_base_limpia_sumativo1.csv", index=False)